library/src/main/java/com/datatorrent/lib/db/jdbc/JDBCDimensionalOutputOperator.java

```diff
@@ -0,0 +1,462 @@
+    sb.append(tableName);
+    sb.append(" (");
+
+    addList(sb, keyNames);
+    sb.append(",");
+    addList(sb, aggregateNames);
+
+    sb.append(") VALUES (");
+
+    for (int qCounter = 0;; qCounter++) {
+      sb.append("?");
+
+      if (qCounter == keyNames.size() + aggregateNames.size() - 1) {
+        break;
+      }
+
+      sb.append(",");
+    }
+
+    sb.append(") ON DUPLICATE KEY UPDATE ");
+
+    addOnDuplicate(sb, aggregateNames);
+
+    return sb.toString();
+  }
+
+  private void addOnDuplicate(StringBuilder sb, List<String> names)
+  {
+    LOG.info("add Duplicate");
+    for (int index = 0;; index++) {
+
+      String name = names.get(index);
+      sb.append(name);
+      sb.append("=");
+      sb.append("VALUES(");
+      sb.append(name);
+      sb.append(")");
+
+      if (index == names.size() - 1) {
+        break;
+      }
+
+      sb.append(",");
+    }
+  }
+
+  private void addList(StringBuilder sb, List<String> names)
+  {
+    for (int index = 0;; index++) {
+      sb.append(names.get(index));
+
+      if (index == names.size() - 1) {
+        break;
+      }
+
+      sb.append(",");
+    }
+  }
+
+  /**
+   * This sets the table names that corresponds to the dimensions combinations
+   * specified in your {@link DimensionalConfigurationSchema}. The structure of
+   * this property is as follows: <br/>
+   * <br/>
+   * <ol>
+   * <li>The first key is the dimension combination id assigned to a dimension
+   * combination in your {@link DimensionalConfigurationSchema}. <br/>
+   * <br/>
+   * The dimensions descriptor id is determined by the following factors:
+   * <ul>
+   * <li>The dimensions combination specified in the
+   * {@link DimensionalConfigurationSchema}.</li>
+   * <li>The the Time Buckets defined in your
+   * {@link DimensionalConfigurationSchema}.</li>
+   * </ul>
+   * The dimensions descriptor id is computed in the following way:
+   * <ol>
+   * <li>The first dimensions descriptor id is 0</li>
+   * <li>A dimension combination is selected</li>
+   * <li>A time bucket is selected</li>
+   * <li>The current dimension combination and time bucket pair is assigned a
+   * dimensions descriptor id</li>
+   * <li>The current dimensions descriptor id is incremented</li>
+   * <li>Steps 3 - 5 are repeated until all the time buckets are done</li>
+   * <li>Steps 2 - 6 are repeated until all the dimension combinations are done.
+   * </li>
+   * </ol>
+   * <br/>
+   * <</li>
+   * <li>The second key is the name of an aggregation being performed for that
+   * dimensions combination.</li>
+   * <li>The value is the name of the output Mysql table</li>
+   * </ol>
+   * 
+   * @param tableNames
+   *          The table names that corresponds to the dimensions combinations
+   *          specified in your {@link DimensionalConfigurationSchema}.
+   */
+  public void setTableNames(Map<Integer, Map<String, String>> tableNames)
+  {
+    this.tableNames = Preconditions.checkNotNull(tableNames);
```

Comment:
Why do we need NotNull check here, there is already NotNull annotation on tableNames.
Does this Preconditions.checkNotNull checks for each entry inside the Map for null values?